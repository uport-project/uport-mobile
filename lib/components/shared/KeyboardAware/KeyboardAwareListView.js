// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import React from 'react'
import PropTypes from 'prop-types'

import { FlatList } from 'react-native'

import KeyboardAwareBase from './KeyboardAwareBase'

export default class KeyboardAwareListView extends KeyboardAwareBase {
  render() {
    const initialOpacity = this.props.startScrolledToBottom ? 0 : 1
    return (
      <FlatList
        {...this.props}
        {...this.style}
        opacity={initialOpacity}
        contentInset={{ bottom: this.state.keyboardHeight }}
        ref={r => {
          this._keyboardAwareView = r
        }}
        onLayout={layoutEvent => {
          this._onKeyboardAwareViewLayout(layoutEvent.nativeEvent.layout)
        }}
        onScroll={event => {
          this._onKeyboardAwareViewScroll(event.nativeEvent.contentOffset)
          if (this.props.onScroll) {
            this.props.onScroll(event)
          }
        }}
        onContentSizeChange={() => {
          this._updateKeyboardAwareViewContentSize()
        }}
        scrollEventThrottle={200}
      />
    )
  }
}

KeyboardAwareListView.propTypes = {
  onScroll: PropTypes.func,
}

KeyboardAwareListView.defaultProps = {
  ...KeyboardAwareBase.defaultProps,
}
