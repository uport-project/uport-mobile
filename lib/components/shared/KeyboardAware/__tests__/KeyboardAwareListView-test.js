// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { FlatList } from 'react-native'
import React from 'react'
import KeyboardAwareListView from '../KeyboardAwareListView'
import renderer from 'react-test-renderer'

describe('KeyboardAwareListView', () => {
  it('renders the KeyboardAwareListView', () => {
    const tree = renderer.create(<KeyboardAwareListView dataSource={{ rowIdentities: [] }} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('calls onKeyboardAwareViewLayout from onLayout', () => {
    this.wrapper = global.shallow(<KeyboardAwareListView dataSource={{ rowIdentities: [] }} />)
    const instance = this.wrapper.instance()
    instance._onKeyboardAwareViewLayout = jest.fn()
    this.wrapper
      .find(FlatList)
      .props()
      .onLayout({ nativeEvent: true })
    expect(instance._onKeyboardAwareViewLayout).toHaveBeenCalled()
  })
  it('calls onKeyboardAwareViewScroll from onScroll', () => {
    const onScroll = jest.fn()
    this.wrapper = global.shallow(<KeyboardAwareListView dataSource={{ rowIdentities: [] }} onScroll={onScroll} />)
    const instance = this.wrapper.instance()
    instance._onKeyboardAwareViewScroll = jest.fn()
    this.wrapper
      .find(FlatList)
      .props()
      .onScroll({ nativeEvent: true })
    expect(instance._onKeyboardAwareViewScroll).toHaveBeenCalled()
    expect(onScroll).toHaveBeenCalled()
  })
  it('calls updateKeyboardAwareViewContentSize from onContentSizeChange', () => {
    this.wrapper = global.shallow(<KeyboardAwareListView dataSource={{ rowIdentities: [] }} />)
    const instance = this.wrapper.instance()
    instance._updateKeyboardAwareViewContentSize = jest.fn()
    this.wrapper
      .find(FlatList)
      .props()
      .onContentSizeChange()
    expect(instance._updateKeyboardAwareViewContentSize).toHaveBeenCalled()
  })
})
