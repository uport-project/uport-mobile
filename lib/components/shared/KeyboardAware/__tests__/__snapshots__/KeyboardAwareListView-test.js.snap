// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardAwareListView renders the KeyboardAwareListView 1`] = `
<RCTScrollView
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  dataSource={
    Object {
      "rowIdentities": Array [],
    }
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  opacity={1}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={200}
  scrollToBottomOnKBShow={false}
  scrollToInputAdditionalOffset={75}
  startScrolledToBottom={false}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;
